src/visitor.h

```diff
@@ -66,6 +75,59 @@ class StopVisitor : public Visitor {
 
 void preorder_traversal_stop(const Basic &b, StopVisitor &v);
 
+template<class T>
+class BaseVisitor : public Visitor {
+
+public:
+    T *p_;
+public:
+    BaseVisitor(T* p) : p_ {p} {
```

Comment:
When is this constructor being called? It seems these two lines bypass it:
```c++
StrPrinter strPrinter;
return strPrinter.apply(*this);
```
So I am confused a bit how this can work at all.